Threat model analysis revision (#2)
oduameh/mithril: docs/website/root/mithril/advanced/threat-model.md

:::danger

This is a draft version of the **Mithril threat model**, prepared by the **Mithril core team**. It is open for external feedback and contributions before being finalized.

> - To contribute, use the **Edit this page** link at the bottom of the page.

Reviewer: ```suggestion
- To contribute, use the **Edit this page** option at the bottom of the page.
```

Author: updated

- If you discover a security vulnerability, report it responsibly by following the [security vulnerability disclosure policy](https://github.com/input-output-hk/mithril/blob/main/SECURITY.md).

:::

This document provides an analysis of various security threats and possible mitigations affecting the Mithril network and its participants. It adopts an adversarial mindset, analyzing the system from an attacker's perspective.
 
The threat model is a living document, updated to reflect the [latest Mithril version](https://github.com/input-output-hk/mithril/releases/latest).

## System analysis

### System description

:::info To do
- Consider reducing these details and moving them into the [architecture page](https://mithril.network/doc/mithril/mithril-network/architecture).
- Also consider updating the [protocol page](https://mithril.network/doc/mithril/mithril-protocol/protocol).
:::

> The system consists of three main components: signers, an aggregators, and clients.

Reviewer: ```suggestion
The system consists of three main components: signers, aggregators, and clients.
```

Author: updated

- **Mithril signers** certify Cardano chain data by using a Mithril signing key. They must have access to both a trusted `cardano-node` and the Mithril signing key to operate.

> - **Mithril signing keys** are rotated every epoch and require certification by the Cardano KES key. Therefore, the Mithril signers must have access to the KES key to register a signing key for each epoch.

Reviewer: ```suggestion
- **Mithril signing keys** are rotated every epoch and require certification by the Cardano KES key. Therefore, Mithril signers must have access to the KES key to register a signing key for each epoch.
```

Author: Updated

#### Deployment architecture

This document focuses on the standard deployment architecture, where a Mithril signer runs alongside the block-producing node, and the aggregator is accessed only through a relay.

[![Mithril - Architecture](mithril-network/images/architecture.svg)](mithril-network/images/architecture.svg)

Additional information is available in:

- [Mithril Network Architecture](https://mithril.network/doc/mithril/mithril-network/architecture)  

> - [Run a Mithril signer as a stake pool operator (SPO)](https://mithril.network/doc/manual/getting-started/run-signer-node)

Reviewer: ```suggestion
- [Mithril network architecture](https://mithril.network/doc/mithril/mithril-network/architecture)  
- [Run a Mithril signer as a stake pool operator (SPO)](https://mithril.network/doc/manual/getting-started/run-signer-node).
```

Author: updated

> ### External Dependencies

Reviewer: ```suggestion
### External dependencies
```

Author: updated

> Below is an example of listing dependencies for the `mithril-signer` component:

Reviewer: More edits are needed to the copy below this sentence:
![image](https://github.com/user-attachments/assets/5b3e8f1e-8591-43ec-b33c-a50a8d1fadb9)

Author: updated

## Assumptions

The core Mithril protocol is considered safe, and its analysis is out of scope for this document. It is computationally infeasible to:

- Forge a valid aggregate signature from forged signing keys  

> - Forge individual signatures to impersonate a legitimate signer  

Reviewer: ```suggestion
- Forge individual signatures to impersonate a legitimate signer.  
```

Author: updated

More information about the core Mithril protocol and its security is available in the [research paper](https://iohk.io/en/research/library/papers/mithril-stake-based-threshold-multisignatures/).

## Assets

> For each asset, we first identify which part of the **CIA Triad** (Confidentiality, Integrity, Availability) it requires. An asset may be a resource, a piece of data (for example, keys), or a process that may be affected. We then identify threats and possible countermeasures.

Reviewer: ```suggestion
For each asset, we first identify which part of the **CIA triad** (Confidentiality, Integrity, Availability) it requires. An asset may be a resource, a piece of data (for example, keys), or a process that may be affected. We then identify threats and possible countermeasures.
```

Author: updated

#### KES private keys

- The KES key is present only on the block-producing (BP) node but needs to be shared with both the `cardano-node` process and the `mithril-signer` process.  
- KES keys are needed by `mithril-signer` to sign a verification key along with an operational certificate that authenticates the key for this stake pool ID.  
- This signing happens at every epoch.

  - **Confidentiality**: Yes  
    Capturing KES private keys allows an attacker to impersonate a registered SPO on-chain and produce blocks on its behalf until the keys are rotated.

  - **Integrity**: Yes  
    Rotating a compromised KES key is a time-consuming process that can be carried out even if the associated KES period has not fully passed. See [KES period docs](https://github.com/input-output-hk/cardano-node-wiki/blob/main/docs/stake-pool-operations/7_KES_period.md).

  - **Availability**: Yes  
    If the KES key is unavailable, signing cannot proceed.

#### Block diffusion

Block diffusion ensures the timely propagation of blocks that are produced locally or received from upstream peers.

- **Confidentiality**: No (block diffusion happens openly on the network)  

> - **Integrity**: Yes (partially—if data is corrupted in transit, it can affect network consistency)  

Reviewer: ```suggestion
- **Integrity**: Yes (partially – if data is corrupted in transit, it can affect network consistency)  
```

Author: updated

> - **Availability**: Yes (failure to diffuse blocks harms a block producer’s economic viability and can disrupt the network)

Reviewer: ```suggestion
- **Availability**: Yes (failure to diffuse blocks harms a block producer’s economic viability and can disrupt the network).
```

Author: updated

#### Block production

Block production is the process of minting new blocks by block producers, using a stake-based random lottery. The `mithril-signer` must run on the same host as a block producer, as it requires access to the KES signing key.

- **Confidentiality**: Yes?  
  A block producer’s schedule is private information; if leaked, adversaries might carry out **grinding attacks** to manipulate the nonce to the producer’s disadvantage.

- **Integrity**: Yes  
  Incorrect or invalid data can hamper block production.

- **Availability**: Yes  
  Block production is critical for SPO revenue; preventing a producer from creating blocks can harm the SPO’s operational capabilities.

> #### Cardano Chain database

Reviewer: ```suggestion
#### Cardano chain database
```

Author: updated

A `cardano-node` maintains an on-disk database containing the chain’s history. This database is updated by the node when new blocks are diffused across the network or minted, and it contains a cache of the ledger state.

> The `mithril-signer` needs access to a *trusted* and *current* chain database in order to sign snapshots.

Reviewer: ```suggestion
The `mithril-signer` needs access to a *trusted* and *current* chain database to sign snapshots.
```

Author: updated

- **Confidentiality**: No (block data is publicly replicated)  
- **Integrity**: Yes  

> - **Availability**: Yes

Reviewer: ```suggestion
- **Availability**: Yes.
```

Author: updated

> #### Cardano Ledger state

Reviewer: ```suggestion
#### Cardano ledger state
```

Author: updated

> Access to an accurate ledger state is needed by the `mithril-signer` to retrieve a reliable *stake distribution*. This currently involves a local connection (direct with Pallas or indirect with `cardano-cli`) to a trusted `cardano-node`. The ledger state and stake distribution are also used by the node to determine leader schedules, meaning any corruption here affects block production.

Reviewer: Turn passive into active: 'The Mihril signer requires access...'

Author: updated

- **Confidentiality**: No
- **Integrity**: Yes (same, inaccurate SD will make key registration and signing process invalid)

> - **Availability**: Yes (without SD, signer cannot register keys nor validly use other signers' keys)

Reviewer: ```suggestion
- **Availability**: Yes (without SD, signer cannot register keys nor validly use other signers' keys).
```

Author: Updated

#### Mithril signing keys

SPOs generate their Mithril signing keys for each epoch to sign snapshots. If attackers obtain these keys, they can impersonate the SPO and produce invalid snapshots. Currently, the signing keys are stored temporarily on disk; they are used for two epochs after creation and are deleted two epochs later. Additionally, these keys are not encrypted.

- **Confidentiality**: Yes (compromise allows an attacker to impersonate the signer)  
- **Integrity**: Yes (an invalid key is useless)  

> - **Availability**: Yes (the signer needs the key at every signing round; unavailability prevents signing)

Reviewer: ```suggestion
- **Availability**: Yes (the signer needs the key at every signing round; unavailability prevents signing).
```

Author: Updated

#### Mithril signers registration

A `mithril-signer` must register a new verification key each epoch with the aggregator (and ultimately other signers).

- **Confidentiality**: No (only public verification keys and proofs of possession are exchanged)  
- **Integrity**: Yes (the key registration must be correct and timely for a given epoch)  

> - **Availability**: Yes (the signer needs aggregator access to register its key)

Reviewer: ```suggestion
- **Availability**: Yes (the signer needs aggregator access to register its key).
```

Author: Updated

#### Mithril signatures diffusion

Mithril signers produce signatures for new immutable files or other relevant data in the chain database. These signatures are derived from a random lottery based on the signer’s stake and protocol parameters.  
Preventing signers from submitting signatures reduces the overall number of signatures, which could allow an attacker a greater influence over snapshot production.

- **Confidentiality**: No  
- **Integrity**: No (signatures are effectively tamper-proof)  

> - **Availability**: Yes

Reviewer: ```suggestion
- **Availability**: Yes.
```

Author: updated

#### Mithril protocol parameters

Protocol parameters coordinate valid multi-signature production and are served by the aggregator.

- **Confidentiality**: No (these parameters must be public)  
- **Integrity**: Yes (tampering with them could lead to invalid signatures)  

> - **Availability**: Yes

Reviewer: ```suggestion
- **Availability**: Yes.
```

Author: Updated

#### Mithril genesis signing key

The genesis signing key is stored in IOG’s VaultWarden and used only once to generate the genesis certificate.

- **Confidentiality**: Yes  
- **Integrity**: Yes (?)  

> - **Availability**: No (not needed unless a re-genesis process is required; a new key could be used in that case)

Reviewer: ```suggestion
- **Availability**: No (not needed unless a re-genesis process is required; a new key could be used in that case).
```

Author: Updated

#### Era configuration files

> The [Mithril Network Upgrade Strategy](https://mithril.network/doc/adr/4) describes how Mithril eras activate features on all nodes at specific epoch boundaries.

Reviewer: ```suggestion
The [Mithril network upgrade strategy](https://mithril.network/doc/adr/4) describes how Mithril eras activate features on all nodes at specific epoch boundaries.
```

Author: Updated

> An era address is used by signers to identify the current Mithril Era, which defines the structure of snapshots and signatures. It is stored in [GitHub](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.addr) and only modifiable via a merged PR.

Reviewer: ```suggestion
An era address is used by signers to identify the current Mithril era, which defines the structure of snapshots and signatures. It is stored in [GitHub](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.addr) and is only modifiable via a merged PR.
```

Author: updated

- **Confidentiality**: No (these are public files)  
- **Integrity**: Yes (tampering could disrupt multi-signature generation)  

> - **Availability**: Yes

Reviewer: ```suggestion
- **Availability**: Yes.
```

Author: updated

#### Era activation

The current and next (if any) eras are announced on-chain with an era activation marker. (See [ADR](https://mithril.network/doc/adr/4#era-activation-marker).)

- **Confidentiality**: No (public)  
- **Integrity**: Yes (tampering the marker could break multi-signature generation)  

> - **Availability**: Yes (the marker is on-chain)

Reviewer: ```suggestion
- **Availability**: Yes (the marker is on-chain).
```

#### Era verification key

> The era verification key is in [GitHub](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.vkey) and only modifiable through a merged PR.

Reviewer: ```suggestion
The era verification key is in [GitHub](https://raw.githubusercontent.com/input-output-hk/mithril/main/mithril-infra/configuration/release-mainnet/era.vkey) and is only modifiable through a merged PR.
```

Author: Updated

- **Confidentiality**: No  
- **Integrity**: No (the protocol enforces integrity checks)  

> - **Availability**: Yes (needed to verify the certificate chain)

Reviewer: ```suggestion
- **Availability**: Yes (needed to verify the certificate chain).
```

Author: Yes

#### Era signing key

This key is stored in IOG’s VaultWarden and used only when a new era is announced or activated.

- **Confidentiality**: Yes  
- **Integrity**: Yes (?)  

> - **Availability**: No 

Reviewer: ```suggestion
- **Availability**: No. 
```

Author: Updated

### Client-side only assets

These assets matter only when downloading and verifying certificates and artifacts (for example, snapshots).

#### Mithril certificate verification process

Mithril clients, such as a CLI tool or the [web-based explorer](https://mithril.network/explorer/), download snapshots and verify their certificates using the `mithril-client` library.

- **Confidentiality**: No  
- **Integrity**: No (snapshots and certificates are assumed to be secure by protocol design)  

> - **Availability**: Yes (?)

Reviewer: ```suggestion
- **Availability**: Yes (?).
```

Author: Updated

#### Mithril certificates

> These are produced by the aggregator from individual signatures. Their security depends on a chain of trust that ultimately references a Genesis certificate.

Reviewer: ```suggestion
These are produced by the aggregator from individual signatures. Their security depends on a chain of trust that ultimately references a genesis certificate.
```

Author: Updated

- **Confidentiality**: No  
- **Integrity**: No (the protocol design ensures integrity)  

> - **Availability**: No (certificates are a fallback; a Cardano node client can retrieve data from the network, albeit more slowly)

Reviewer: ```suggestion
- **Availability**: No (certificates are a fallback; a Cardano node client can retrieve data from the network, albeit more slowly).
```

Author: updated

#### Mithril artifacts

Mithril artifacts are generated by the aggregator for specific data types (for example, a compressed snapshot of a Cardano node database). The aggregator hosts these artifacts for client retrieval.

- **Confidentiality**: No  
- **Integrity**: No (they are verified against their certificate)  

> - **Availability**: No (a fallback exists via the original Cardano network)

Reviewer: ```suggestion
- **Availability**: No (a fallback exists via the original Cardano network).
```

Author: updated

#### Mithril genesis verification key

The genesis verification key is stored in [GitHub](https://github.com/input-output-hk/mithril/blob/main/mithril-infra/configuration/release-mainnet/genesis.vkey) and is only modifiable through an approved PR.

- **Confidentiality**: No  
- **Integrity**: No (the protocol enforces integrity)  

> - **Availability**: Yes (needed to verify the entire certificate chain)

Reviewer: ```suggestion
- **Availability**: Yes (needed to verify the entire certificate chain).
```

Author: updated

> ## Threat & Mitigations

Reviewer: ```suggestion
## Threat and mitigations
```

Author: updated

:::info

- This list of threat and mitigations is not exhaustive.

> - [Developers portal](https://developers.cardano.org/docs/operate-a-stake-pool/hardening-server) already provides thorough documentation on how to harden a linux-based host to run `cardano-node`.

Reviewer: ```suggestion
- [Developer Portal](https://developers.cardano.org/docs/operate-a-stake-pool/hardening-server) already provides thorough documentation on hardening a Linux-based host to run `cardano-node`.
```

Author: updated

:::

### Resource exhaustion on Cardano relay

A denial-of-service (DoS) attack targeting a `mithril-signer` running alongside a `cardano-node` acting as a relay.
- **Assets at risk**:
  - [Block diffusion](#block-diffusion)

>   - [Mithril signatures diffusion](#mithril-signatures-diffusion)

Reviewer: ```suggestion
  - [Mithril signatures diffusion](#mithril-signatures-diffusion).
```

Author: updated

#### Block production exhaustion

- A malfunctioning or malicious `mithril-signer` could starve a `cardano-node` of resources, preventing it from producing and diffusing blocks in time.  
- Compromising a block-producing node severely harms an SPO’s economic viability.

### Resource exhaustion on Mithril aggregator

A DoS on the `mithril-aggregator`.
- **Assets at risk**:
  - [Mithril signers registration](#mithril-signers-registration)
  - [Mithril signatures diffusion](#mithril-signatures-diffusion)
  - [Mithril certificates](#mithril-certificates)

>   - [Mithril artifacts](#mithril-artifacts)

Reviewer: ```suggestion
  - [Mithril artifacts](#mithril-artifacts).
```

Author: updated

### Integrity of the Cardano block producer database

Data integrity of the Cardano block producer’s on-disk database could be compromised either by action of the Mithril signer or by an attacker with access to the signer.
- **Assets at risk**:
  - [Block production](#block-production)

>   - [Cardano Chain database](#cardano-chain-database)

Reviewer: ```suggestion
  - [Cardano chain database](#cardano-chain-database).
```

Author: updated

> - **Mitigation**: Assign the Mithril signer only read-only permissions to the Cardano block producer’s database folder.

Reviewer: ```suggestion
- **Mitigation**: assign the Mithril signer *read-only* permissions to the Cardano block producer’s database folder.
```

Author: updated

> ## References

Reviewer: The copy below needs edits to ensure 'Sentence case titles':

![image](https://github.com/user-attachments/assets/0039b997-12cd-4683-b82c-ea573540f50a)

Author: updated
